### PR TITLE
fix MM swap by adding approve

### PR DESCRIPTION
**Pull request by mandalorian-101**

### CODE REVIEW

File: contracts/Strategy.sol [hunk 1]
```diff
@@ -254,17 +257,18 @@ contract Strategy is BaseStrategy {
                 (uint256 _mToken, ) = MMFarmingPool(mmFarmingPool).userInfo(mmFarmingPoolId, address(this));
                 require(_mToken >= _mvGap, '!insufficientMTokenInMasterChef');
```
Comment (Grandthrax) [inline contracts/Strategy.sol]: can this be made to always revert by someone? perhaps by airdroppig in some tokens to confuse calculations?

Comment (Grandthrax) [reply contracts/Strategy.sol]: Also i think this ruins your liquidateall solution. it will revert if we ask for more than our assets

File: contracts/Strategy.sol [hunk 2]
```diff
@@ -254,17 +257,18 @@ contract Strategy is BaseStrategy {
...
+            MMVault(mmVault).withdraw(liquidateAll? IERC20(mmVault).balanceOf(address(this)) : _mShare);
             uint256 _after = IERC20(want).balanceOf(address(this));
             require(_after > _before, '!mismatchMushroomsVaultWithdraw');
 
-            _liquidatedAmount = _after.sub(_before);
+            _liquidatedAmount = _after;
             _loss = _amountNeeded > _liquidatedAmount? _amountNeeded.sub(_liquidatedAmount): 0; // due to possible withdrawal fee
-            return (_liquidatedAmount, _loss);		
+            require(_liquidatedAmount + _loss <= _amountNeeded);
+            return (_liquidatedAmount, liquidateAll? 0 : _loss);
```
Comment (Grandthrax) [inline contracts/Strategy.sol]: why is loss 0 if you liquidated all?

Comment (Grandthrax) [reply contracts/Strategy.sol]: i think loss should be difference between vault.strategies(address(this)).totalDebt and_liquidatedAmount. You have withdrawn all and we still have debt left that is loss.

**Review by Grandthrax — COMMENTED** [review]
not sure liquidate all is working correctly. please review my comments

File: contracts/Strategy.sol (hunk 1 above)
Comment (mushroomsforest) [reply contracts/Strategy.sol]: There is no way to airdrop to Mushrooms farming pool to affect the `userInfo` output for this strategy. If the attacker want to airdrop `want` token or `mmVault` share token directly to strategy, I do not see the economic incentive here because what is airdropped will be taken as strategy asset to get returned to yVault

File: contracts/Strategy.sol (hunk 2 above)
Comment (mushroomsforest) [reply contracts/Strategy.sol]: PR updated to make this `liquidateAll` logic more clear. Thanks for sharing the insight here
